clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -2869,9 +2870,23 @@ static void setLinkageForGV(llvm::GlobalValue *GV, const NamedDecl *ND) {
     GV->setLinkage(llvm::GlobalValue::ExternalWeakLinkage);
 }
 
+static bool hasExistingGeneralizedTypeMD(llvm::Function *F) {
+  llvm::MDNode *MD = F->getMetadata(llvm::LLVMContext::MD_type);
+  if (!MD)
+    return false;
+  return MD->hasGeneralizedMDString();
+}
+
 void CodeGenModule::createFunctionTypeMetadataForIcall(const FunctionDecl *FD,
                                                        llvm::Function *F) {
-  // Only if we are checking indirect calls.
+  if (CodeGenOpts.CallGraphSection && !hasExistingGeneralizedTypeMD(F) &&
+      (!F->hasLocalLinkage() ||
+       F->getFunction().hasAddressTaken(nullptr, /*IgnoreCallbackUses=*/true,
+                                        /*IgnoreAssumeLikeCalls=*/true,
+                                        /*IgnoreLLVMUsed=*/false)))
+    F->addTypeMetadata(0, CreateMetadataIdentifierGeneralized(FD->getType()));
```

Comment:
Relying on the IR function properties feels wrong here, can you detect whether this is indirect from the source context? I'm also not sure IgnoreCallbackUses is correct here (IgnoreAssumeLikeCalls is also potentially questionable)

Reply:
Doesn't this function just emit the type metadata if the function *could* be used as an indirect call target? The function name doesn't really communicate that though (at least to me). 

Based on that reading, I'd say limiting it to any non-local function (e.g. could be addr taken in another TU) and any local function that has its address taken makes sense for the call-graph use-case.

Or was there some other aspect that concerns you?

Reply:
I also agree about the IgnoreCallbacks and IgnoreAssumeLIkeCalls.

Reply:
If clang is emitting the initial IR, then won't any IR based use query be unreliable? Later uses could yet to be determined. But also it shouldn't be fundamentally problematic to be conservatively correct and emit the annotation?

Reply:
Well, yeah, I guess that would be true, and you're right that just always doing it would be correct.  Its also closer to what's done for CFI, so we should probably just do that to start, and we can try to limit the amount of type metadata later if we find its unnecessary/safe to do so.

Reply:
I think this still needs to be addressed...